src/crystal/hasher.cr

```diff
@@ -0,0 +1,59 @@
+require "crystal/system/random"
+
+# :nodoc:
+struct Crystal::Hasher
+  # Implementation of a Hasher to compute a fast and safe hash
+  # value for primitive and basic Crystal objects. All other
+  # hashes are computed based on these.
+  #
+  # TODO: the implementation is naive and should probably use
+  # another algorithm like SipHash or FNV.
+
+  @@seed = uninitialized UInt64
+  Crystal::System::Random.random_bytes(Slice.new(pointerof(@@seed).as(UInt8*), 8))
+
+  property value : UInt64 = @@seed
+
+  def nil
+    self
+  end
+
+  def bool(value)
+    (value ? 1 : 0).hash(self)
+  end
+
+  def int(value)
+    @value = @value * 31 + value.to_u64
+    self
+  end
+
+  def float(value)
+    @value = @value * 31 + value.to_f64.unsafe_as(UInt64)
```

Comment:
I think we should at use [this](https://llvm.org/docs/LangRef.html#llvm-canonicalize-intrinsic) on the floats.

Reply:
We can leave it for a next PR

Reply:
TODO then?

Reply:
There's a big todo at the top of the struct. On fact in this PR I could have used zero as a seed, but I wanted to see that it worked with it.

Reply:
I know what you are asking is not exactly covered by the above TODO. Still, I'd like to tackle these issues one at a time. After we merge this, I'll probably wait for someone (you? :-)) to use float normalization, though I'll require a test to see why it's needed. Then another PR can tackle another (better) hash algorithm, etc.

Reply:
`canonicalize` is OK, but implementation proposed by @funny-falcon (PyHash inspired) is more robust and generic normalization that takes in care canonicalization too. Will propose it as separate PR.